### PR TITLE
#1018 Migration: Append_GetRange

### LINKED ISSUE #1018: Command Migration: ('APPEND', 'GETRANGE')
This issue tracks the migration of the mentioned commands - ('APPEND', 'GETRANGE') to make them compatible across the three protocols supported by the Dice database: RESP, HTTP, and WebSocket.
DiceDB now supports more than one protocols (Resp/http/websocket). We don't want eval operation of each command to be strictly bound with any of the protocols. Currently eval function return statements of each command is bounded to the RESP protocol. 

*The migration is required because:*

- The current implementation is specific to the RESP protocol
- Evaluation functions of each command must be refactored to ensure protocol-agnostic behavior.
- Evaluation functions should return raw responses instead of protocol specific responses.

The goal is to make the command logic protocol-independent, allowing all three protocols to call the same core functionality seamlessly.

### Requirements

- Refactor the evaluation function of `[Command Name]` to be generic.
- Ensure the evaluation function does not include protocol-specific logic.

### Migration Steps

1. **Analyze Current Implementation**
   - Review the current codebase to understand how the command logic is implemented.

2. **Refactor Return Logic**
   - Create a new function with the function definition as `evalXXX(args []string, store *dstore.Store) *EvalResponse` under the file `/internal/eval/store_eval.go`
   - Analyse return statements of the eval and modify them to send raw types without encoding
   - Use errors in the return statements from `/errors/migrated_errors.go` file
   - Use perdefined responses from `/internal/clientio/resp.go` file as RespType variables

3. **Command/Worker specific Changes**
   - Make `IsMigrated` flag to true in the commands information under `/internal/eval/commands.go` file.
   - Use the newly written eval function against the `NewEval` parameter in the same command structure.
   - Delete the old eval function from the `/internal/eval/eval.go` file.
   - Add the migrated command to the `/internal/worker/CommandsMeta` map and make it's type as `SingleShard`

4. **Update Unit Tests**
   - Refactor existing unit tests to accommodate the new implementation.
   - Add new unit tests if necessary to cover all possible cases.

5. **Integration Tests**
   - Run all integration tests to ensure successful migration.
   - Ensure that each protocol (RESP, HTTP, WebSocket) works correctly after migration.

### Checklist

- [ ] Migrated the evalXXX function with the latest definition
- [ ] Update or add unit tests for the new implementation.
- [ ] All unit tests pass successfully.
- [ ] Ensure all integration tests pass successfully.

### Additional Notes

- Describe any edge cases that need to be handled in the generic function.
- Mention any protocol-specific optimizations required in wrappers.
  
If there are any questions or concerns about this migration, please mention them here.

### Related Issues/PRs

- Sample implementation for the Get, Set, GetSet and SetEx can be found in [this Pull Request.](https://github.com/DiceDB/dice/pull/766)

### PR DESCRIPTION
**Pull request by bhima2001**
migrated GETRANGE from eval.go to store_eval.go, modified all the test cases. This pr closes #1018

 - [x] Migrated the evalXXX function with the latest definition
 - [x] Update or add unit tests for the new implementation.
 - [x] All unit tests pass successfully.
 - [x] Ensure all integration tests pass successfully.
 - [x] Ensure integration tests are added for the migrated command on multi-threaded resp server.
 - [x] Move Integration tests for the respective commands under the RESP integration tests directory from Async directory
 - [x] Please validate that the documentation for the respective commands is up to date. If not then consider adding them.

### CODE REVIEW

File: internal/eval/eval_test.go
Comment (benjaminmishra) [inline internal/eval/eval_test.go:4117]: We have a function called `runMigratedEvalTests` that does this same thing and can be reused here.

**Review by benjaminmishra — CHANGES_REQUESTED** [review]

**Review by benjaminmishra — APPROVED** [review]

**Comment by bhima2001** [comment]
I need to migrate APPEND too.

**Comment by soumya-codes** [comment]
@bhima2001 lets ensure that the corresponding integration tests are migrated as well. cc @AshwinKul28

**Comment by JyotinderSingh** [comment]
Please fix the merge conflicts

**Comment by bhima2001** [comment]
@JyotinderSingh @benjaminmishra I am not able to identify the reason for this error. can you please nudge into the right direction. Thanks.

**Comment by benjaminmishra** [comment]
@bhima2001 I can see at least one test failing may be more. Can I pull you branch to check it out ?

**Comment by bhima2001** [comment]
You are talking about  this test case right "FAIL: TestGetSet/GETSET_error_when_key_exists_but_does_not_hold_a_string_value". But this testcase is not failing when i testing it individually/

File: integration_tests/commands/async/getrange_test.go
Comment (lucifercr07) [inline integration_tests/commands/async/getrange_test.go:1]: We can remove these tests now as we've migrated.

**Review by lucifercr07 — CHANGES_REQUESTED** [review]

File: integration_tests/commands/async/getrange_test.go
Comment (lucifercr07) [inline integration_tests/commands/async/getrange_test.go:1]: This also needs to be removed, since migrated.

**Review by lucifercr07 — APPROVED** [review]
LGTM, minor comment.

File: integration_tests/commands/async/getrange_test.go
Comment (bhima2001) [reply integration_tests/commands/async/getrange_test.go:1]: [b3a09a1](https://github.com/DiceDB/dice/pull/1095/commits/b3a09a10077bea807b2b9dd3e1f42042b477885e)

**Comment by soumya-codes** [comment]
The changes look great @bhima2001.

@apoorvyadav1111 has added a document-template/sample-document for the command documentation.

Please refer to the following files:
https://github.com/DiceDB/dice/blob/master/docs/command_docs_template.md
https://github.com/DiceDB/dice/blob/master/docs/sample_command_docs.md

We need to align our documentation around this. This will greatly help us be production ready.
Can you please ensure this is the case.

**Comment by bhima2001** [comment]
Hey, @soumya-codes I have added docs for both  "GETRANGE" and "APPEND"  commands using the mentioned format. Please let me know if any further changes need to be made.

File: internal/eval/commands.go
Comment (apoorvyadav1111) [inline internal/eval/commands.go:992]:
```suggestion
		Arity:      3,
```

File: internal/eval/commands.go
Comment (apoorvyadav1111) [inline internal/eval/commands.go:1020]:
```suggestion
		Arity:      2,
```

File: internal/eval/store_eval.go
Comment (apoorvyadav1111) [inline internal/eval/store_eval.go:339]: it looks like we will return "" even if the start and end values are invalid. This is a deviation from the expected behavior for `GETRANGE`. Could you please check on this and add this test case in integration and unit tests. Thanks

File: internal/eval/store_eval.go
Comment (apoorvyadav1111) [inline internal/eval/store_eval.go:374]: getrange works for other data type such as bytearray as well, please check and update code accordingly. Please add test cases in integration tests and unit tests. Thanks

File: internal/eval/store_eval.go
Comment (apoorvyadav1111) [inline internal/eval/store_eval.go:583]: append can also work with other data types such as byte array, please check for others

File: internal/eval/store_eval.go
Comment (apoorvyadav1111) [inline internal/eval/store_eval.go:608]: same as above

File: internal/eval/store_eval.go
Comment (apoorvyadav1111) [inline internal/eval/store_eval.go:619]: please look into this #1037 and coordinate with the assignee on how to proceed. We are aiming to get migration of commands with production ready code and if a fix is known we would like to get it fixed as soon as possible. If you guys can work together on this, this would be great.

**Review by apoorvyadav1111 — CHANGES_REQUESTED** [review]
Hi @bhima2001 , Thanks for the changes, Please consider my review comments.

File: internal/eval/store_eval.go
Comment (bhima2001) [reply internal/eval/store_eval.go:339]: If the start and end  are invalid then we throw the following error "ERR value is not an integer or out of range". The above mention case is when an object with the given key is not present in the store

**Comment by apoorvyadav1111** [comment]
Hi @bhima2001 , please resolve the conversations that have been addressed and let me know if its ready for review.

File: docs/src/content/docs/commands/APPEND.md
Comment (bhima2001) [reply docs/src/content/docs/commands/APPEND.md]: [5ec96be](https://github.com/DiceDB/dice/pull/1095/commits/5ec96beeff2cbe27ae4dc6e9271e94ccc3169c35)

File: docs/src/content/docs/commands/GETRANGE.MD
Comment (bhima2001) [reply docs/src/content/docs/commands/GETRANGE.MD]: [5ec96be](https://github.com/DiceDB/dice/pull/1095/commits/5ec96beeff2cbe27ae4dc6e9271e94ccc3169c35)

File: internal/eval/commands.go
Comment (bhima2001) [reply internal/eval/commands.go]: [263dbf3](https://github.com/DiceDB/dice/pull/1095/commits/263dbf3f1ecbf4198e2166c7e7454036f60e033a)

File: internal/eval/commands.go
Comment (bhima2001) [reply internal/eval/commands.go]: [263dbf3](https://github.com/DiceDB/dice/pull/1095/commits/263dbf3f1ecbf4198e2166c7e7454036f60e033a)

File: integration_tests/commands/websocket/getrange_test.go
Comment (bhima2001) [reply integration_tests/commands/websocket/getrange_test.go:1]: [c0171ff](https://github.com/DiceDB/dice/pull/1095/commits/c0171ffb14004c450fc31743356e881ec3ca31dd)

File: integration_tests/commands/websocket/append_test.go
Comment (bhima2001) [reply integration_tests/commands/websocket/append_test.go:59]: [c0171ff](https://github.com/DiceDB/dice/pull/1095/commits/c0171ffb14004c450fc31743356e881ec3ca31dd)

File: integration_tests/commands/resp/getrange_test.go
Comment (bhima2001) [reply integration_tests/commands/resp/getrange_test.go:1]: [c0171ff](https://github.com/DiceDB/dice/pull/1095/commits/c0171ffb14004c450fc31743356e881ec3ca31dd)

File: integration_tests/commands/resp/append_test.go
Comment (bhima2001) [reply integration_tests/commands/resp/append_test.go]: [c0171ff](https://github.com/DiceDB/dice/pull/1095/commits/c0171ffb14004c450fc31743356e881ec3ca31dd)

File: integration_tests/commands/http/append_test.go
Comment (bhima2001) [reply integration_tests/commands/http/append_test.go]: [c0171ff](https://github.com/DiceDB/dice/pull/1095/commits/c0171ffb14004c450fc31743356e881ec3ca31dd) --> for all the tests related changes

File: docs/src/content/docs/commands/GETRANGE.MD
Comment (lucifercr07) [inline docs/src/content/docs/commands/GETRANGE.MD:5]: Shall we draft it as?
`The GETRANGE command in DiceDB is used to get a substring of a string, provided the start and end indices`

File: docs/src/content/docs/commands/GETRANGE.MD
Comment (lucifercr07) [inline docs/src/content/docs/commands/GETRANGE.MD:16]: Shall this be?
`The name of the key containing the string.`

File: docs/src/content/docs/commands/GETRANGE.MD
Comment (lucifercr07) [inline docs/src/content/docs/commands/GETRANGE.MD:17]: `substring` -> `substring.`

File: docs/src/content/docs/commands/GETRANGE.MD
Comment (lucifercr07) [inline docs/src/content/docs/commands/GETRANGE.MD:23]: `stirng` -> `string`

File: docs/src/content/docs/commands/GETRANGE.MD
Comment (lucifercr07) [inline docs/src/content/docs/commands/GETRANGE.MD:32]: Shall we change it to?
`Both start and end can be negative, which removes characters from the string, starting from the end + 1 position on the right.`

File: docs/src/content/docs/commands/GETRANGE.MD
Comment (lucifercr07) [inline docs/src/content/docs/commands/GETRANGE.MD:46]: `Invalid input type for start and end`

File: docs/src/content/docs/commands/GETRANGE.MD
Comment (lucifercr07) [inline docs/src/content/docs/commands/GETRANGE.MD:96]: `Trying to use `GETRANGE` on an invalid data type.`

**Review by lucifercr07 — CHANGES_REQUESTED** [review]
LGTM, Minor doc changes suggested.

File: internal/eval/store_eval.go
Comment (apoorvyadav1111) [reply internal/eval/store_eval.go:339]: Yes, but even in that case, we must not return a result but an error

**Comment by bhima2001** [comment]
@apoorvyadav1111  I would suggest to create a new issue for functionality changes for getrange and assign it to me. Because there are a lot of changes in this PR already and resolving merge conflicts everyday is getting more and more difficult. I would love to pick the code changes issue. 
Thanks.

**Comment by apoorvyadav1111** [comment]
Hi @bhima2001, please create an issue with the functionalities we discussed and I will assign that to you. Please close that issue on priority. This PR is ready to merge once all comments from @lucifercr07 are addressed. 

Thanks for taking this humongous task and closing it.

**Comment by apoorvyadav1111** [comment]
Please reach out and tag me on discord once all comments are addressed and rebased. Will ensure we don't have to rebase again.

**Review by apoorvyadav1111 — CHANGES_REQUESTED** [review]
Please rebase your PR. Changes are approved but cannot merge until conflicts

File: docs/src/content/docs/commands/GETRANGE.MD
Comment (apoorvyadav1111) [inline docs/src/content/docs/commands/GETRANGE.MD:16]:
```suggestion
| `key`     | The name of the key containing the string.                                           | String  | Yes      |
```

File: docs/src/content/docs/commands/GETRANGE.MD
Comment (apoorvyadav1111) [inline docs/src/content/docs/commands/GETRANGE.MD:96]:
```suggestion
Trying to use `GETRANGE` on an invalid data type.
```

**Review by apoorvyadav1111 — CHANGES_REQUESTED** [review]
Hi @bhima2001 , Thank you these changes and addressing all the comments Approving the PR with an additional commit to fix some typos + addressing review from @lucifercr07 .

Thanks

**Review by apoorvyadav1111 — APPROVED** [review]

**Pull request merged**
